src/parser.ml

```diff
@@ -989,19 +989,58 @@ let link_label allow_balanced_brackets st =
   in
   loop 0 false
 
+(* based on https://erratique.ch/software/uucp/doc/Uucp/Case/index.html#caselesseq *)
 let normalize s =
-  let buf = Buffer.create (String.length s) in
-  let rec loop start seen_ws i =
-    if i >= String.length s then Buffer.contents buf
-    else
-      match s.[i] with
-      | ' ' | '\t' | '\010' .. '\013' -> loop start true (succ i)
-      | _ as c ->
-          if (not start) && seen_ws then Buffer.add_char buf ' ';
-          Buffer.add_char buf (Char.lowercase_ascii c);
-          loop false false (succ i)
+  let canonical_caseless_key s =
+    let b = Buffer.create (String.length s * 2) in
+    let to_nfd_and_utf_8 =
+      let n = Uunf.create `NFD in
+      let rec add v =
+        match Uunf.add n v with
+        | `Await | `End -> ()
+        | `Uchar u ->
+            Uutf.Buffer.add_utf_8 b u;
+            add `Await
+      in
+      add
+    in
+    let add =
+      let n = Uunf.create `NFD in
+      let rec add v =
+        match Uunf.add n v with
+        | `Await | `End -> ()
+        | `Uchar u ->
+            (match Uucp.Case.Fold.fold u with
+            | `Self -> to_nfd_and_utf_8 (`Uchar u)
+            | `Uchars us -> List.iter (fun u -> to_nfd_and_utf_8 (`Uchar u)) us);
+            add `Await
+      in
+      add
```

Comment:
With all the `add`s flying around here, I am prone to get a bit confused. Perhaps it would help clarify for readers like me to give this outer `add` a more specific name? Perhaps like `add_nfd`?